pkg/cvo/updatepayload.go

```diff
@@ -290,20 +280,24 @@ func (r *payloadRetriever) pruneJobs(ctx context.Context, retain int) error {
 
 // copyPayloadCmd returns a shell command that copies CVO and release manifests from the default location
 // to the target dir.
-// It is made up of 2 commands:
-// `mkdir -p <target dir> && mv <default cvo manifest dir> <target cvo manifests dir>`
-// `mkdir -p <target dir> && mv <default release manifest dir> <target release manifests dir>`
 func copyPayloadCmd(tdir string) string {
-	var (
-		fromCVOPath = filepath.Join(payload.DefaultPayloadDir, payload.CVOManifestDir)
-		toCVOPath   = filepath.Join(tdir, payload.CVOManifestDir)
-		cvoCmd      = fmt.Sprintf("mkdir -p %s && mv %s %s", tdir, fromCVOPath, toCVOPath)
+	baseDir, targetName := filepath.Split(tdir)
+	tmpDir := filepath.Join(baseDir, fmt.Sprintf("%s-%s", targetName, randutil.String(5)))
 
-		fromReleasePath = filepath.Join(payload.DefaultPayloadDir, payload.ReleaseManifestDir)
-		toReleasePath   = filepath.Join(tdir, payload.ReleaseManifestDir)
-		releaseCmd      = fmt.Sprintf("mkdir -p %s && mv %s %s", tdir, fromReleasePath, toReleasePath)
-	)
-	return fmt.Sprintf("%s && %s", cvoCmd, releaseCmd)
+	cleanupCmd := fmt.Sprintf("rm -fR %s", filepath.Join(baseDir, "*"))
+
+	tmpDirCmd := fmt.Sprintf("mkdir %s", tmpDir)
+
+	fromCVOPath := filepath.Join(payload.DefaultPayloadDir, payload.CVOManifestDir)
+	toCVOPath := filepath.Join(tmpDir, payload.CVOManifestDir)
+	cvoCmd := fmt.Sprintf("mv %s %s", fromCVOPath, toCVOPath)
+
+	fromReleasePath := filepath.Join(payload.DefaultPayloadDir, payload.ReleaseManifestDir)
+	toReleasePath := filepath.Join(tmpDir, payload.ReleaseManifestDir)
+	releaseCmd := fmt.Sprintf("mv %s %s", fromReleasePath, toReleasePath)
+
+	moveInPlaceCmd := fmt.Sprintf("mv %s %s", tmpDirCmd, tdir)
```

Comment:
I think you may have to first create `tdir` to make the `rename` atomic, https://man7.org/linux/man-pages/man2/rename.2.html

Reply:
From [POSIX][1]:

> This `rename()` function is equivalent for regular files to that defined by the ISO C standard. Its inclusion here expands that definition to include actions on directories and specifies behavior when the new parameter names a file that already exists. That specification requires that the action of the function be atomic.

The "If `newpath` already exists ..." section of the Linux man page is also in the "... and also when the target does exist..." category.  Although it would be nice if someone said "this is atomic when the target does not exist" more explicitly on either side.

[1]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html

Reply:
[This article][1] talks through the atomic rename approach, and from that, it looks like we might need a pre-rename `fsync` to be very secure about the transaction.  I'm not sure that covering the kernel-crash corner case is worth the overhead of calling `fsync`.

[1]: https://lwn.net/Articles/789600/

Reply:
I've pushed 4be737e0 -> a5af89d3 with an updated commit message to walk through this.